assets/javascripts/discourse/components/chat-toggle-membership-btn.js

```diff
@@ -0,0 +1,91 @@
+import Component from "@ember/component";
+import I18n from "I18n";
+import { inject as service } from "@ember/service";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+import ChatApi from "discourse/plugins/discourse-chat/discourse/lib/chat-api";
+import { action, computed } from "@ember/object";
+
+export default class ChatToggleMembershipBtn extends Component {
+  @service chat;
+  tagName = "";
+
+  channel = null;
+  afterToggle = null;
+  isLoading = false;
+  labelType = null;
+  joinTitle = null;
+  joinIcon = null;
+  joinClasses = "";
+  leaveTitle = null;
+  leaveIcon = null;
+  leaveClasses = "";
+
+  @computed("channel.following")
+  get label() {
+    if (this.labelType === "none") {
+      return "";
+    }
+
+    if (this.labelType === "short") {
+      if (this.channel.following) {
+        return I18n.t("chat.channel_settings.leave");
+      } else {
+        return I18n.t("chat.channel_settings.join");
+      }
+    }
+
+    if (this.channel.following) {
+      return I18n.t("chat.channel_settings.leave_channel");
+    } else {
+      return I18n.t("chat.channel_settings.join_channel");
+    }
+  }
+
+  @action
+  onJoinChannel() {
+    this.set("isLoading", true);
+
+    return ChatApi.followChatChannel(this.channel.id)
```

Comment:
Maybe we implement some `channel.updateMembership(membership)` function in channel model, and call it in chat-api client that way this update membership part is seamless.

Reply:
So something like this inside the `channel` model? Then called like `this.channel.updateMembership(false/true).then(() => this.onToggle.?())` ?

```javascript
updateMembership(follow) {
  if (follow) {
    return ChatApi.followChatChannel(this.channel.id)
      .then((membership) => {
        this.channel.setProperties({
          following: true,
          muted: membership.muted,
          desktop_notification_level: membership.desktop_notification_level,
          mobile_notification_level: membership.mobile_notification_level,
          memberships_count: membership.user_count,
        });

        this.chat.startTrackingChannel(this.channel);
      });
  } else {
    return ChatApi.unfollowChatChannel(this.channel.id)
      .then((membership) => {
        this.channel.setProperties({
          expanded: false,
          following: false,
          memberships_count: membership.user_count,
        });

        this.chat.stopTrackingChannel(this.channel);
      })
      .catch(popupAjaxError);
  }
}
```

Reply:
Ohhh I see, yes sounds good :+1: